contracts/liquidityMining/KyberFairLaunch.sol

```diff
@@ -0,0 +1,469 @@
+    require(_rewardLockBps <= BPS, 'update: invalid lock bps');
+    require(_rewardPerBlock > 0, 'update: invalid reward per block');
+
+    (
+      pool.endBlock,
+      pool.rewardLockBps,
+      pool.rewardPerBlock
+    ) = (
+      _endBlock,
+      _rewardLockBps,
+      _rewardPerBlock
+    );
+    emit UpdatePool(_pid, _endBlock, _rewardLockBps, _rewardPerBlock);
+  }
+
+  /**
+  * @dev deposit to tokens to accumulate rewards
+  * @param _pid: id of the pool
+  * @param _amount: amount of stakeToken to be deposited
+  */
+  function deposit(uint256 _pid, uint256 _amount) external override {
+    PoolInfo storage pool = poolInfo[_pid];
+    UserInfo storage user = userInfo[_pid][msg.sender];
+    require(pool.stakeToken != address(0), 'deposit: not accept deposit');
+    updatePoolRewards(_pid);
+    if (user.amount > 0) _harvest(msg.sender, _pid);
+    IERC20Ext(pool.stakeToken).safeTransferFrom(msg.sender, address(this), _amount);
+    user.amount = user.amount.add(_amount);
+    user.rewardDebt = user.amount.mul(pool.accRewardPerShare).div(1e12);
+    emit Deposit(msg.sender, _pid, block.number, _amount);
+  }
+
+  /**
+  * @dev withdraw token (of the sender) from pool, also harvest reward
+  * @param _pid: id of the pool
+  * @param _amount: amount of stakeToken to withdraw
+  */
+  function withdraw(uint256 _pid, uint256 _amount) external override {
+    _withdraw(_pid, _amount);
+  }
+
+  /**
+  * @dev withdraw all tokens (of the sender) from pool, also harvest reward
+  * @param _pid: id of the pool
+  */
+  function withdrawAll(uint256 _pid) external override {
+    _withdraw(_pid, userInfo[_pid][msg.sender].amount);
+  }
+
+  /**
+  * @dev emergency withdrawal function to allow withdraw all deposited token (of the sender)
+  *   without harvesting the reward
+  * @param _pid: id of the pool
+  */
+  function emergencyWithdraw(uint256 _pid) external override {
+    PoolInfo storage pool = poolInfo[_pid];
+    UserInfo storage user = userInfo[_pid][msg.sender];
+    uint256 amount = user.amount;
+    user.amount = 0;
+    user.rewardDebt = 0;
+    if (amount > 0) {
+      IERC20Ext(pool.stakeToken).safeTransfer(msg.sender, user.amount);
+    }
+    emit EmergencyWithdraw(msg.sender, _pid, block.number, user.amount);
+  }
+
+  /**
+  * @dev harvest reward from pool for the sender
+  * @param _pid: id of the pool
+  */
+  function harvest(uint256 _pid) external override {
+    PoolInfo storage pool = poolInfo[_pid];
+    UserInfo storage user = userInfo[_pid][msg.sender];
+    updatePoolRewards(_pid);
+    _harvest(msg.sender, _pid);
+    user.rewardDebt = user.amount.mul(pool.accRewardPerShare).div(1e12);
+  }
+
+  /**
+  * @dev harvest rewards from all pools for the sender
+  */
+  function harvestAll() external override {
+    uint256 length = poolInfo.length;
+    for(uint256 i = 0; i < length; i++) {
+      _harvest(msg.sender, i);
```

Comment:
not updating user.rewardDebt?

Reply:
Updated rewards for pool and user's debt